alibi/explainers/similarity/base.py

```diff
@@ -0,0 +1,171 @@
+from abc import ABC
+from typing import TYPE_CHECKING, Callable, Union, Tuple, Any
+
+import numpy as np
+from tqdm import tqdm
+
+from alibi.api.interfaces import Explainer
+from alibi.explainers.similarity.backends import _select_backend
+
+
+if TYPE_CHECKING:
+    import tensorflow
+    import torch
+
+
+class BaseSimilarityExplainer(Explainer, ABC):
+    """Base class for similarity explainers."""
+
+    def __init__(self,
+                 predictor: 'Union[tensorflow.keras.Model, torch.nn.Module]',
+                 loss_fn: '''Union[Callable[[tensorflow.Tensor, tensorflow.Tensor], tensorflow.Tensor],
+                                   Callable[[torch.Tensor, torch.Tensor], torch.Tensor]]''',
+                 sim_fn: Callable[[np.ndarray, np.ndarray], np.ndarray],
+                 store_grads: bool = False,
+                 seed: int = 0,
+                 backend: str = "tensorflow",
+                 device: str = "cpu",
```

Comment:
Consider those as literals?

Reply:
I would use enums again since this is not public, we have one for `Framework` in `utils.frameworks`, we could also create a `Device` one. In general I would say public interface -> `Literal`, private interface -> a `str` `Enum` with the same values.